src/mod_extdisco.erl

```diff
@@ -0,0 +1,132 @@
+
+-define(EXT_DISCO, <<"urn:xmpp:extdisco:2">>).
+
+-spec start(jid:server(), list()) -> ok.
+start(Host, Opts) ->
+    IQDisc = gen_mod:get_opt(iqdisc, Opts, one_queue),
+    mod_disco:register_feature(Host, ?EXT_DISCO),
+    gen_iq_handler:add_iq_handler(ejabberd_local, Host,
+        ?EXT_DISCO, ?MODULE, process_iq, IQDisc).
+
+-spec stop(jid:server()) -> ok.
+stop(Host) ->
+    mod_disco:unregister_feature(Host, ?EXT_DISCO),
+    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?EXT_DISCO).
+
+-spec process_iq(jid:jid(), jid:jid(), mongoose_acc:t(), jlib:iq()) ->
+    {mongoose_acc:t(), jlib:iq()}.
+process_iq(_From, _To, Acc, #iq{type = get, sub_el = SubEl} = IQ) ->
+    Response = case request_type(SubEl) of
+        all_services ->
+            RequestedServices = get_external_services(),
+            create_iq_response(RequestedServices);
+        {credentials, {Type, Host}} ->
+            Services = get_external_services([Type], []),
+            RequestedServices = lists:filter(fun({_, Opts}) ->
+                gen_mod:get_opt(host, Opts, undefined) == binary_to_list(Host)
+                end, Services),
+            create_iq_response_credentials(RequestedServices);
+        {selected_services, Type} ->
+            RequestedServices = get_external_services([Type], []),
+            create_iq_response_services(RequestedServices, Type);
+        _ ->
+            []
+    end,
+    {Acc, IQ#iq{type = result, sub_el = Response}}.
+
+request_type(#xmlel{name = <<"services">>} = Element) ->
+    case exml_query:attr(Element, <<"type">>) of
+        undefined -> all_services;
+        Services -> {selected_services, binary_to_atom(Services, utf8)}
+    end;
+request_type(#xmlel{name = <<"credentials">>, children = [C]}) ->
+    Host = exml_query:attr(C, <<"host">>),
+    Type = exml_query:attr(C, <<"type">>),
+    {credentials, {binary_to_atom(Type, utf8), Host}};
+request_type(_) ->
+    {error, unknown_request}.
+
+create_iq_response(Services) ->
+    #xmlel{name = <<"services">>,
+        attrs = [{<<"xmlns">>, ?EXT_DISCO}],
+        children = prepare_services_element(Services, [])}.
+
+create_iq_response_services(Services, Type) ->
+    #xmlel{name = <<"services">>,
+        attrs = [{<<"xmlns">>, ?EXT_DISCO}, {<<"type">>, atom_to_binary(Type, utf8)}],
+        children = prepare_services_element(Services, [])}.
+
+create_iq_response_credentials(Services) ->
+    #xmlel{name = <<"credentials">>,
+        attrs = [{<<"xmlns">>, ?EXT_DISCO}],
+        children = prepare_services_element(Services, [])}.
+
+get_external_services() ->
+    get_external_services([stun, turn], []).
+get_external_services([], ServicesWithOpts) ->
+    ServicesWithOpts;
+get_external_services([Type | Types], ServicesWithOpts) ->
+    case gen_mod:get_module_opt(?MYNAME, ?MODULE, Type, []) of
+        [] -> get_external_services(Types, ServicesWithOpts);
+        Opts -> get_external_services(Types, [{Type, Opts} | ServicesWithOpts])
+    end.
```

Comment:
Here we're making `stun` and `turn` the defaults, but what if we've actually configured more services to be discovered? In that case we wouldn't be returning to the client "all" the available services 😕
I think that `gen_mod` should be getting all that is configured and greedily return that.
This could also help skipping the conversions to and from binary, perhaps, but I haven't really given this last line a deep thought.

Reply:
I think sticking to some known types of external services that are supported would be safer to implement. The goal of this implementation is to allow to pass on the information related to STUN/TURN servers. In the XEP I could only find FTP as another external service type example, so maybe we could add that as well. If there would come up a need to support discovery of other services we could easily add them, but in general I would be against passing on whatever we have in the config. The boundaries and limitations for the configuration of this feature will be described in the documentation.

Reply:
The thing is that I think the code should be fully functional from the get-go, and not to have to be modified every time something new pops up, even if the modification is tiny, it can be an unnecessary overhead 😕 Usually the config is a trusted source, only operators touch it, so I think it can be trusted (?)

Reply:
Ok, will modify the code accordingly.